File: 0048-rotate-image/0048-rotate-image.js
/**
 * @param {number[][]} matrix
 * @return {void} Do not return anything, modify matrix in-place instead.
 */
var rotate = function(matrix) {
    let array = [];
    
    for (let i = 0 ; i < matrix.length; i++) {
        for (let j = 0; j < matrix.length; j++) {
            array.push(matrix[i][j]);
        }
    }
    
    for (let i = 0 ; i < matrix.length; i++) {
        for (let j = 0; j < matrix.length; j++) {
            matrix[j][i] = array.shift();
        }
    }
    
    for (let i = 0 ; i < matrix.length; i++) {
       matrix[i] = matrix[i].reverse();
    }

    return matrix;
};

File: 0048-rotate-image/0048-rotate-image.py
class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        map = {}

        # transpose matrix
        for i in range(len(matrix)):
            for j in range(len(matrix[0])):
                map[(i, j)] = matrix[i][j]

                if (j, i) not in map:
                    matrix[i][j] = matrix[j][i]
                else:
                    matrix[i][j] = map[(j, i)]

        # reverse matrix rows
        for row in matrix:
            row.reverse()

